api/xds_protocol.rst

```diff
@@ -36,43 +36,122 @@ occurs.
+2. Incremental xDS: incremental, separate gRPC stream for each resource type
+3. Aggregated Discovery Service (ADS): SotW, aggregate stream for all resource types
+4. Incremental ADS: incremental, aggregate stream for all resource types
+
+In general, new xDS implementations are encouraged to use the last variant.
+
+RPC Services and Methods for Each Variant
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For the non-aggregated API variants, there is a separate RPC service for each resource type. Each of these RPC services can
+provide a method for each of the SotW and Incremental API variants. Here are the RPC services and methods for each resource
+type:
+
+-  Listener: Listener Discovery Service (LDS)
+   -  SotW: ListenerDiscoveryService.StreamListeners
+   -  Incremental: ListenerDiscoveryService.DeltaListeners
+-  RouteConfiguration: Route Discovery Service (RDS)
+   -  SotW: RouteDiscoveryService.StreamRoutes
+   -  Incremental: RouteDiscoveryService.DeltaRoutes
+-  VirtualHost: Virtual Host Discovery Service (VHDS)
+   -  SotW: N/A
+   -  Incremental: VirtualHostDiscoveryService.DeltaVirtualHosts
+-  Cluster: Cluster Discovery Service (CDS)
+   -  SotW: ClusterDiscoveryService.StreamClusters
+   -  Incremental: ClusterDiscoveryService.DeltaClusters
+-  ClusterLoadAssignment: Endpoint Discovery Service (EDS)
+   -  SotW: EndpointDiscoveryService.StreamEndpoints
+   -  Incremental: EndpointDiscoveryService.DeltaEndpoints
+-  Secret: Secret Discovery Service (SDS)
+   -  SotW: SecretDiscoveryService.StreamSecrets
+   -  Incremental: SecretDiscoveryService.DeltaSecrets
+-  Runtime: Runtime Discovery Service (RTDS)
+   -  SotW: RuntimeDiscoveryService.StreamRuntime
+   -  Incremental: RuntimeDiscoveryService.DeltaRuntime
+
+In the aggregated API variants, all resource types are multiplexed on a single gRPC stream, where each resource type is
+treated as a separate logical stream within the aggregated stream. In effect, it simply combines all of the above separate
+APIs into a single stream by treating requests and responses for each resource type as a separate sub-stream on the single
+aggregated stream. The RPC service and methods for the aggregated API variants are:
+
+-  SotW: AggregatedDiscoveryService.StreamAggregatedResources
+-  Incremental: AggregatedDiscoveryService.DeltaAggregatedResources
+
+For all of the SotW methods, the request type is :ref:`DiscoveryRequest <envoy_api_msg_DiscoveryRequest>` and the response
+type is :ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>`.
+
+For all of the incremental methods, the request type is :ref:`DeltaDiscoveryRequest <envoy_api_msg_DeltaDiscoveryRequest>` and
+the response type is :ref:`DeltaDiscoveryResponse <envoy_api_msg_DeltaDiscoveryResponse>`.
+
+Configuring Which Variant to Use
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In Envoy, a :ref:`ConfigSource <envoy_api_msg_core.ConfigSource>` can be specified independently for each xDS API,
```

Comment:
Why is this only in Envoy? The `ConfigSource` concept is plumbed across the entire API, so anyone using xDS needs to be aware of this.

Reply:
It's not really plumbed across the whole API in a consistent way.  As you may recall, you and @mattklein123 didn't want to move forward with #8200, which would have added a link to CDS from the RDS response.

The reason I said that this is only in Envoy is that gRPC does not use the ConfigSource proto in its bootstrap file.  We support only ADS, and our bootstrap file contains only the name of the ADS server and specifies the credentials to use to talk to that server.  Because we do not have the link from RDS to CDS in the API, the fact that we support only ADS is convenient, because it avoids the need for our bootstrap file to separately configure the server to use for CDS data.  But this basically means that we have no way to configure getting CDS data from somewhere different from where we get LDS data.  In the future, as we add more support for direct federation, I think we will still need something like #8200 to address this.

In terms of the doc, what I was trying to do here was to keep this focused on the xDS API rather than on the details of the ways that Envoy or gRPC are configured.  But I guess the two are not easily separable.

I've added a new "API flow" section above, which described the differences between Envoy and gRPC.  And I've rewritten this section to clarify the differences in how these variants are configured for both clients.

Please let me know what you think.